### PR TITLE
[ML] Data Frame HLRC Preview API

### PR DESCRIPTION
**Pull request by davidkyle**

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java [hunk 1]
```diff
@@ -77,9 +77,9 @@ public DataFrameTransformConfig(final String id,
                                     final String dest,
                                     final QueryConfig queryConfig,
                                     final PivotConfig pivotConfig) {
-        this.id = Objects.requireNonNull(id);
+        this.id = id;
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java]: The way I see this being used by the client is that they already have a crafted `DataFrameTransformConfig` that they want to preview. Consequently, these fields should still be required. 

The main reason for not requiring them on the backend of this call is so the UI (or others) can preview a not (full) valid config as it is being built out. I am not sure this is necessary or even preferential for the HLRC.

**Review by benwtrent — COMMENTED** [review]
One comment about allowing `dest` and `id` to be null.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java (hunk 1 above)
Comment (davidkyle) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java]: If an Id is set then preview errors with:

```
reason=Inconsistent id; 'faafa' specified in the body differs from 'transform-preview' specified as a URL argument]];
```

Which is why I allowed null but now I realise I was just patching over a bug. The `Id` field should be overwritten the same way destination is overwritten here:

https://github.com/elastic/elasticsearch/blob/e3f2f05e6c4ac76e92bf723c8fc3247d4849faef/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/action/PreviewDataFrameTransformAction.java#L62

I pushed a commit fixing that. 

These `Objects.requireNonNull()` checks are in the wrong place logically they should be in `PutDataFrameTransformRequest.validate()` which is neater and follows the expected framework usage. Preview validation would be different allowing null values. The server then performs any more complicated validation. I'll make that change in this pr.

Comment (davidkyle) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java]: I pushed https://github.com/elastic/elasticsearch/pull/40206/commits/c05775fedde170a1290387ba0f98d7ee781e3e64 for proper validation on put and preview

**Review by benwtrent — APPROVED** [review]
🖖 live long and validate

**Pull request merged**